vm/src/stdlib/zlib.rs

```diff
@@ -0,0 +1,150 @@
+use crate::function::OptionalArg;
+use crate::obj::{objbytes::PyBytesRef, objint::PyIntRef};
+use crate::pyobject::{create_type, ItemProtocol, PyObjectRef, PyResult};
+use crate::vm::VirtualMachine;
+
+use adler32::RollingAdler32 as Adler32;
+use crc32fast::Hasher as Crc32;
+use flate2::{write::ZlibEncoder, Compression, Decompress, FlushDecompress, Status};
+use libz_sys as libz;
+use num_traits::cast::ToPrimitive;
+
+use std::io::Write;
+
+// copied from zlibmodule.c (commit 530f506ac91338)
+const MAX_WBITS: u8 = 15;
+const DEF_BUF_SIZE: usize = 16 * 1024;
+
+pub fn make_module(vm: &VirtualMachine) -> PyObjectRef {
+    let ctx = &vm.ctx;
+
+    let zlib_error = create_type("error", &ctx.type_type, &ctx.exceptions.exception_type);
+
+    py_module!(vm, "zlib", {
+        "crc32" => ctx.new_rustfunc(zlib_crc32),
+        "adler32" => ctx.new_rustfunc(zlib_adler32),
+        "compress" => ctx.new_rustfunc(zlib_compress),
+        "decompress" => ctx.new_rustfunc(zlib_decompress),
+        "error" => zlib_error,
+        "Z_DEFAULT_COMPRESSION" => ctx.new_int(libz::Z_DEFAULT_COMPRESSION),
+        "Z_NO_COMPRESSION" => ctx.new_int(libz::Z_NO_COMPRESSION),
+        "Z_BEST_SPEED" => ctx.new_int(libz::Z_BEST_SPEED),
+        "Z_BEST_COMPRESSION" => ctx.new_int(libz::Z_BEST_COMPRESSION),
+        "DEF_BUF_SIZE" => ctx.new_int(DEF_BUF_SIZE),
+        "MAX_WBITS" => ctx.new_int(MAX_WBITS),
+    })
+}
+
+macro_rules! checksum_fn {
+    ($fn_name: ident, $h_new: expr, $h_update: expr, $h_compute: expr, $begin_state: expr) => {
+        fn $fn_name(
+            data: PyBytesRef,
+            begin_state: OptionalArg<PyIntRef>,
+            vm: &VirtualMachine,
+        ) -> PyResult<PyObjectRef> {
+            let data = data.get_value();
+
+            let begin_state = begin_state
+                .into_option()
+                .as_ref()
+                .map(|v| v.as_bigint().to_i32().unwrap())
+                .unwrap_or($begin_state);
+
+            let mut hasher = $h_new(begin_state as u32);
+            $h_update(&mut hasher, &data);
+
+            let checksum: u32 = $h_compute(hasher);
+
+            Ok(vm.new_int(checksum))
+        }
+    };
+}
+
+checksum_fn!(
```

Comment:
The macro is used twice, so I would suggest to simply write out the two methods for readability. It's about the same amount on lines, but is somewhat easier to read. What do you think?

Reply:
Ok, at the first time i had done it with ```PyFuncArgs``` and ```arg_check!``` - it was much longer then. I'll rewrite it

Reply:
ok, done! :smiley: